File: src/main/java/javabot/operations/RFCOperation.java
package javabot.operations;

import com.antwerkz.sofia.Sofia;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.inject.Singleton;
import javabot.Message;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.pircbotx.Channel;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

/**
 * Displays RFC url and title
 */
@Singleton
public class RFCOperation extends BotOperation {
    public static final String prefix = "rfc ";
    LoadingCache<String, String> rfcTitleCache = CacheBuilder.newBuilder()
                                                             .maximumSize(100)
                                                             .expireAfterWrite(1, TimeUnit.HOURS)
                                                             .recordStats()
                                                             .build(
                                                                       new CacheLoader<String, String>() {
                                                                           @SuppressWarnings("NullableProblems")
                                                                           public String load(final String url)
                                                                               throws IOException {
                                                                               Document doc = Jsoup.connect(url).get();
                                                                               return doc.title();
                                                                           }
                                                                       });

    @Override
    public boolean handleMessage(final Message event) {
        final String message = event.getValue().toLowerCase();
        if (message.startsWith(prefix)) {
            final String rfcText = message.substring(prefix.length()).trim();
            try {
                int rfc = Integer.parseInt(rfcText);
                try {
                    String url = String.format("http://www.faqs.org/rfcs/rfc%d.html", rfc);
                    getBot().postMessageToChannel(event, Sofia.rfcSucceed(url, rfcTitleCache.get(url)));
                } catch (ExecutionException e) {
                    // from rfc.fail
                    getBot().postMessageToChannel(event, Sofia.rfcFail(rfcText));
                }
            } catch (NumberFormatException e) {
                getBot().postMessageToChannel(event, Sofia.rfcInvalid(rfcText));
            }
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/javabot/operations/RFCOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import com.google.inject.Inject
import com.google.inject.Singleton
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import org.jsoup.HttpStatusException
import org.jsoup.Jsoup
import java.io.IOException
import java.util.Locale

/**
 * Displays RFC url and title
 */
@Singleton
class RFCOperation @Inject constructor(bot: Javabot, adminDao: AdminDao) : BotOperation(bot, adminDao) {
    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        val message = event.value.lowercase(Locale.getDefault())
        if (message.startsWith(PREFIX)) {
            val rfcText = message.substring(PREFIX.length).trim()
            val parts = rfcText.split(Regex("\\s+"))
            val rfcNumPart = parts.getOrNull(0)
            if (rfcNumPart == null || rfcNumPart == "") {
                responses.add(Message(event, Sofia.rfcMissing()))
            } else {
                val rfcNum = rfcNumPart.toIntOrNull()
                if (rfcNum == null) {
                    responses.add(Message(event, Sofia.rfcInvalid(rfcText)))
                } else {
                    try {
                        val anchorType = parts.getOrNull(1)
                        val name = parts.getOrNull(2)
                        val data = if (name != null && ("section".equals(anchorType, true) || "page".equals(anchorType, true))) {
                            load(rfcNum, "#$anchorType-$name")
                        } else {
                            load(rfcNum)
                        }
                        responses.add(Message(event, Sofia.rfcSucceed(data.first, data.second)))
                    } catch (e: HttpStatusException) {
                        responses.add(Message(event, Sofia.rfcFail(rfcText)))
                    } catch (e: IOException) {
                        responses.add(Message(event, Sofia.rfcFail(rfcText)))
                    }
                }
            }
        }
        return responses
    }

    fun load(rfc: Int, anchor: String = ""): Pair<String, String> {
        val url = "https://tools.ietf.org/html/rfc$rfc$anchor"
        val doc = Jsoup.connect(url).get()
        val meta= doc
            .getElementsByTag("span")
            .first { it.className().equals("h1", true) }
            ?.text()
        return Pair(url, meta ?: doc.title())
    }

    companion object {
        val PREFIX: String = "rfc"
    }
}
